examples/mcp-client/src/server.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  Agent,
+  routeAgentRequest,
+  type AgentNamespace,
+  type Connection,
+} from "agents";
+
+type Env = {
+  MCP_SERVERS: { host: string; info: { name: string; version: string } }[];
+  MyAgent: AgentNamespace<MyAgent>;
+};
+
+export class MyAgent extends Agent<Env> {
+  constructor(
+    public ctx: DurableObjectStorage,
+    public env: Env
+  ) {
+    super(ctx, env);
+  }
+
+  async onStart(): Promise<void> {
+    await Promise.all(
+      this.env.MCP_SERVERS.map((server) => {
```

Comment:
Interesting that in this example the agent goes and opens connections to MCP servers when it starts, rather than on-demand. I see what you mean about need for tools caching eventually. Because at inference time need to know what tools are even available.

Reply:
Ideally if the MCP server is well behaved and supports `listChanged` we never need to do caching and can assume our tools/resources/prompts list are always up to date. But for cases where `listChanged` isn't supported, we'd need to do some caching